### PR TITLE
Add MRC file writing support

### LINKED ISSUE #108: Add support for writing MRC files
Following up on #100, I would like to write MRC files, as well as read them.

### PR DESCRIPTION
**Pull request by Pradyumn-cloud**
Fix Issue - #108

Add MRC file writing support
- Implement MRC.write() method 
- Add Grid.export() integration for MRC format 
- Added Test 
- Preserve header information and handle edge cases

### CODE REVIEW

File: gridData/mrc.py [hunk 1]
```diff
@@ -148,5 +148,106 @@ def histogramdd(self):
...
+        
+        Notes
+        -----
+        The data array should be in xyz order (axis 0=X, axis 1=Y, axis 2=Z).
+        
+        If the MRC object was created by reading an existing file, the original
+        header information (including mapc, mapr, maps ordering) is preserved.
+        Otherwise, standard ordering (mapc=1, mapr=2, maps=3) is used.
+        
+        .. versionadded:: 0.8.0
```
Comment (orbeckst) [inline gridData/mrc.py]: change to 1.1.0 (next minor release)

File: gridData/mrc.py [hunk 2]
```diff
@@ -148,5 +148,106 @@ def histogramdd(self):
...
+            h = None
+        
+        # Reverse the transformation done in read()
+        transpose_order = np.argsort(axes_order[::-1])
+        inverse_transpose_order = np.argsort(transpose_order)
+        
+        # Transform our xyz array back to the file's native ordering
+        data_for_file = np.transpose(self.array, axes=inverse_transpose_order)
+        
+        # Ensure proper data type (float32 is standard for mode 2)
```
Comment (orbeckst) [inline gridData/mrc.py]: What's mode 2?

**Review by orbeckst — CHANGES_REQUESTED** [review]
Looks good at first glance but I am not an MRC expert. I'll see if I can get someone else to also have a look.

In the meantime, please also
- ~~update your branch with the latest changes on master~~ (just did that)
- add yourself to AUTHORS
- add an entry to CHANGELOG
- check your coverage and add tests where necessary

File: gridData/core.py [hunk 3]
```diff
@@ -676,6 +679,41 @@ def _export_dx(self, filename, type=None, typequote='"', **kwargs):
...
+        mrc_file.delta = numpy.diag(self.delta)
+        mrc_file.origin = self.origin
+        mrc_file.rank = 3
+        
+        # Transfer header if it exists (preserves axis ordering and other metadata)
+        if hasattr(self, '_mrc_header'):
+            mrc_file.header = self._mrc_header
+        
+        # Write to file
+        mrc_file.write(filename)
```
Comment (marinegor) [inline gridData/core.py:726]: I feel like this should be replaced with `with mrcfile.new(filename) as mrc: ...` block, no? Generally writing to files without using context manager is bad practice.

File: gridData/mrc.py [hunk 4]
```diff
@@ -148,5 +148,106 @@ def histogramdd(self):
...
+        
+        .. versionadded:: 1.1.0
+        """
+        if filename is not None:
+            self.filename = filename
+        
+        # Preserve header if it exists, otherwise use defaults
+        if hasattr(self, 'header'):
+            # File was read - preserve original ordering
+            h = self.header
```
Comment (marinegor) [inline gridData/mrc.py:187]: why is it `self.header` and not `self._mrc_header` as in `core.py`?

**Review by marinegor — DISMISSED** [review]
Hi, thanks for the PR!

I've added couple of comments regarding `self.header` vs `self._mrc_header`, and regarding using `with open(...)`.

Also, perhaps it's a good idea to add a small test MRC (say, 4x4x4 with random numbers) file with/without a header that tests things? This way we can actually track regressions if something changes in `mrcfile` itself, while a roundtrip test won't check that.

File: CHANGELOG [hunk 5]
```diff
@@ -13,6 +13,17 @@ The rules for this file:
...
+  * 1.1.0
+
+  Changes
+
+  * Added MRC file writing support (Issue #108)
+  * Implemented MRC.write() method with proper coordinate transformations
+  * Added Grid.export() integration for MRC format
+  * Preserved header information (mapc/mapr/maps, nstart, cell dimensions)
+  * Added comprehensive tests
+
```
Comment (orbeckst) [inline CHANGELOG]: Too much information, focus on user-facing information, just
```suggestion
  * Added MRC file writing support (Issue #108)
```
is sufficient.

File: CHANGELOG [hunk 6]
```diff
@@ -13,6 +13,17 @@ The rules for this file:
   * accompany each entry with github issue/PR number (Issue #xyz)
 
 ------------------------------------------------------------------------------
+12/12/2025 Pradyumn-cloud
+  * 1.1.0
+
+  Changes
+
```
Comment (orbeckst) [inline CHANGELOG]: All of your changes should go into the existing 1.1.0 section below. Leave the date as ??/??/???? and just
- add your handle after orbeckst
- add the one line "Added MRC file writing support (Issue #108)" in the section *Enhancements*

File: gridData/core.py [hunk 7]
```diff
@@ -676,6 +679,41 @@ def _export_dx(self, filename, type=None, typequote='"', **kwargs):
...
+            Additional keyword arguments (currently ignored)
+        
+        Notes
+        -----
+        * Only orthorhombic unit cells are supported
+        * If the Grid was loaded from an MRC file, the original header
+          information (including axis ordering) is preserved
+        * For new grids, standard ordering (mapc=1, mapr=2, maps=3) is used
+        
+        .. versionadded:: 0.8.0
```
Comment (orbeckst) [inline gridData/core.py]: change to 1.1.0

File: gridData/mrc.py [hunk 8]
```diff
@@ -148,5 +148,106 @@ def histogramdd(self):
...
+            Output filename for the MRC file
+        
+        Notes
+        -----
+        The data array should be in xyz order (axis 0=X, axis 1=Y, axis 2=Z).
+        
+        If the MRC object was created by reading an existing file, the original
+        header information (including mapc, mapr, maps ordering) is preserved.
+        Otherwise, standard ordering (mapc=1, mapr=2, maps=3) is used.
+
```
Comment (orbeckst) [inline gridData/mrc.py:178]: possibly insert a second empty line; check the rendered docs to see of it's correctly rendered

**Review by orbeckst — CHANGES_REQUESTED** [review]
Some minor fixes required and please address @marinegor 's comments. That should also help with test coverage.

File: gridData/core.py (hunk 3 above)
Comment (Pradyumn-cloud) [reply gridData/core.py:726]: `_export_mrc()` method uses `MRC.write() `which properly uses mrcfile.new( filename)
context manager is used -> in mrc.py line (194)

File: gridData/mrc.py (hunk 4 above)
Comment (Pradyumn-cloud) [reply gridData/mrc.py:187]: core stores headers as `_mrc_header `. When exporting, we assign mrc_file.header = self._mrc_header,  giving the MRC instance a header attribute. MRC.write() then finds self.header and uses it. its like different class different name.

Comment (Pradyumn-cloud) [reply gridData/mrc.py:187]: i will provide a test mrc as you mention in some time.

File: gridData/mrc.py (hunk 8 above)
Comment (Pradyumn-cloud) [reply gridData/mrc.py:178]: I tried to do but i am not sure of it

Comment (orbeckst) [reply gridData/mrc.py:178]: Looks good https://griddataformats--147.org.readthedocs.build/en/147/gridData/formats/mrc.html#gridData.mrc.MRC.write

File: gridData/core.py (hunk 3 above)
Comment (orbeckst) [reply gridData/core.py:726]: makes sense

File: gridData/mrc.py (hunk 4 above)
Comment (orbeckst) [reply gridData/mrc.py:187]: Good question and thanks for the explanation. It's a bit confusing but the naming makes sense: the MRC object always has a header but the Grid object does not and we just keep specific mrc metadata around in a special attribute.

**Review by orbeckst — APPROVED** [review]
Sorry it took me a while to get back to the PR. Thank you for addressing all remaining issues and improving the tests. This is good to go and I'll merge.

**Comment by orbeckst** [comment]
@marinegor could you please have a quick look and confirm that all your concerns have been addressed? Thanks!

**Comment by orbeckst** [comment]
@marinegor , to my reading, all your comments have been addressed. I'll wait until tomorrow if you have any comments or want to explicitly approve but I also understand that you're quite busy and might not have the time to come back, in which case I'll dismiss your review and merge.

**Comment by orbeckst** [comment]
Thank you for your contribution @Pradyumn-cloud ! 🎉
